File: 01_arrays_and_hashing/05_top_k_freqeunt_elements/solution.py
from typing import List
from collections import defaultdict, Counter


class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        freq = defaultdict(list)

        for key, ct in Counter(nums).items():
            freq[ct].append(key)

        res = []
        for times in reversed(range(len(nums) + 1)):
            res.extend(freq[times])

            if len(res) >= k:
                return res[:k]

        return res[:k]

File: 01_arrays_and_hashing/05_top_k_freqeunt_elements/solution.go
package main

import (
	"fmt"
)

func topKFrequent(nums []int, k int) []int {
	hashset := make(map[int]int)

	for _, num := range nums {
		hashset[num]++
	}

	reverseDic := map[int][]int{}
	for k, v := range hashset {
		reverseDic[v] = append(reverseDic[v], k)
	}

	result := []int{}
	for i := len(nums); len(result) < k; i-- {
		result = append(result, reverseDic[i]...)
	}

	return result
}

func main() {
	nums := []int{1, 1, 1, 2, 2, 3}
	k := 2
	result := topKFrequent(nums, k)
	fmt.Println(result)
}
